feat: implement full support for mssql dest bulk load

subodh1810 opened the pull request:
Issue : https://github.com/airbytehq/airbyte-internal-issues/issues/11553

jdpgrailsdev commented on airbyte-cdk/bulk/toolkits/load-azure-blob-storage/src/main/kotlin/io/airbyte/cdk/load/file/azureBlobStorage/AzureBlobStreamingUpload.kt:
```diff
@@ -94,4 +96,22 @@ class AzureBlobStreamingUpload(
+    private fun isValidKey(key: String): Boolean {
+        if (key.isBlank() || key.startsWith(RESERVED_PREFIX)) return false
+        if (invalidCharsRegex.containsMatchIn(key)) return false
+        return key.all { it.code in 32..126 }
```
Nit: add comment here about the character range (e.g. it is all ASCII characters, etc)

jdpgrailsdev commented on airbyte-cdk/bulk/toolkits/load-azure-blob-storage/src/main/kotlin/io/airbyte/cdk/load/file/azureBlobStorage/AzureBlobStreamingUpload.kt:
```diff
@@ -94,4 +96,22 @@ class AzureBlobStreamingUpload(
+
+    private fun isValidValue(value: String): Boolean {
+        if (invalidCharsRegex.containsMatchIn(value)) return false
+        return value.all { it.code in 32..126 }
```
Same comment as above

jdpgrailsdev commented on airbyte-integrations/connectors/destination-mssql-v2/src/test/kotlin/io/airbyte/integrations/destination/mssql/v2/MSSQLFormatFileCreatorTest.kt:
```diff
@@ -0,0 +1,213 @@
+import io.airbyte.integrations.destination.mssql.v2.MSSQLFormatFileCreator
+import io.airbyte.integrations.destination.mssql.v2.MSSQLFormatFileCreator.ColumnInfo
+import io.airbyte.integrations.destination.mssql.v2.MSSQLFormatFileCreator.CsvToDbColumn
+import io.mockk.*
```
Nit: remove * imports in favor of explicit ones.

johnny-schmidt commented on airbyte-cdk/bulk/toolkits/load-object-storage/src/main/kotlin/io/airbyte/cdk/load/file/object_storage/ObjectStorageFormattingWriter.kt:
```diff
@@ -99,6 +101,26 @@ class JsonFormattingWriter(
     }
 }
 
+class MSSQLCSVFormattingWriter(
```
What's the purpose of breaking out this separate writer?

Is there a reason you can't just provide the configuration for CSV + root-level flattening.

subodh1810 replied:
Initially that was my approach but because I have to mutate certain values before writing them to the csv file and there is no easy way to re-use the existing writer without having to refactor a bunch of other stuff. (see [comment](https://github.com/airbytehq/airbyte/pull/54195#discussion_r1969155325))

johnny-schmidt commented on airbyte-cdk/bulk/toolkits/load-csv/src/main/kotlin/io/airbyte/cdk/load/data/csv/AirbyteValueToCsvRow.kt:
```diff
@@ -21,25 +21,36 @@ import io.airbyte.cdk.load.data.UnknownValue
+        is TimestampWithoutTimezoneValue -> value
+        is BooleanValue ->
+            when (format) {
+                CsvFormat.MsSql -> if (value) 1 else 0
```
Is there a reason you can't perform this conversion upstream of here?

subodh1810 replied:
am not sure I follow what you mean by `upstream`. I could not find an easy way to apply the conversion in the existing flow and thus had to move it out.

johnny-schmidt commented on airbyte-cdk/bulk/toolkits/load-csv/src/main/kotlin/io/airbyte/cdk/load/data/csv/MSSQLCsvRowValidator.kt:
````diff
@@ -0,0 +1,262 @@
+ *                             Only enable if strict data validation is required.
+ * ```
+ */
+class MSSQLCsvRowValidator(private val validateValuesPreLoad: Boolean) {
````
This feels a little backward. Since we control the conversions, can't we just validate their correctness with tests? And if we know there are any legal record values that can't be serlialized, just null them out with a mapper before converting to CSV?

subodh1810 replied:
So this is to handle the scenarios such as (listing few examples)
1. The column type is number (big decimal) but the precision of value is more than the allowed precision of the database.
2. The column type is non string but the value is string so to make sure that the string value can be converted correctly to the column type, if not replace it with null. 
3. MSSQL represents boolean as bits (0 and 1), so need to convert true->1 and false->0.
4. etc.


Also we are doing this validation/conversion before writing the records to the csv file.

gosusnp commented on airbyte-integrations/connectors/destination-mssql-v2/src/main/kotlin/io/airbyte/integrations/destination/mssql/v2/MSSQLBulkLoadHandler.kt:
```diff
@@ -0,0 +1,244 @@
+            """
+            SELECT TOP 0 *
+            INTO $tempTableName
+            FROM $schemaName.$mainTableName
```
is the tablename escaped here?

subodh1810 replied:
fixed

gosusnp commented on airbyte-integrations/connectors/destination-mssql-v2/src/main/kotlin/io/airbyte/integrations/destination/mssql/v2/MSSQLBulkLoadHandler.kt:
```diff
@@ -0,0 +1,244 @@
+    private fun quoteIdentifier(identifier: String): String {
+        // Split on '.', then bracket each piece (like database.schema.table ->
+        // [database].[schema].[table])
+        return identifier.split(".").joinToString(".") { segment ->
```
feels dangerous to split on `"."` for quoting, I suspect this could break for identifiers with a `.`.

gosusnp commented on airbyte-integrations/connectors/destination-mssql-v2/src/main/kotlin/io/airbyte/integrations/destination/mssql/v2/MSSQLBulkLoadHandler.kt:
```diff
@@ -0,0 +1,244 @@
+        val sourceColumnsCsv = allColumns.joinToString(", ") { "Source.[$it]" }
+
+        return """
+        MERGE INTO $quotedTableName AS Target
```
The part that I don't like here is that the merge query is duplicated, same for column handling and escaping. I suspect this somehow doubles the maintenance effort.

gosusnp commented on airbyte-integrations/connectors/destination-mssql-v2/src/main/kotlin/io/airbyte/integrations/destination/mssql/v2/MSSQLFormatFileCreator.kt:
```diff
@@ -0,0 +1,232 @@
+    private fun pickBcpTypeAndLength(dataType: String, charMaxLength: Int?): Pair<String, Int> {
+        // For CSV-based imports, it's common to treat all columns as text (SQLCHAR).
+        // The length is a best-guess upper bound in bytes.
+        return when (dataType.lowercase()) {
```
Any chance we could align this better on Mssql types?
My concern is that implementations may drift and it could be easier for us to cause unnecessary breaking changes between the two insert modes. My intuition is that a user that started on sql inserts (simple setup) and moves to bulk (for performance) should not have to deal with breaking changes.

subodh1810 replied:
I think since the table creation logic is the same, the final data type of the table would be the same for both the loads. This is specific mapping of sql data type to format file data type. I dont think moving from one sync mode to another would cause this to break.

gosusnp commented on airbyte-integrations/connectors/destination-mssql-v2/src/main/kotlin/io/airbyte/integrations/destination/mssql/v2/MSSQLQueryBuilder.kt:
```diff
@@ -181,7 +180,7 @@ const val COUNT_FROM = """
 
 class MSSQLQueryBuilder(
-    config: MSSQLConfiguration,
+    defaultSchema: String,
```
why are we stringifying here?

subodh1810 replied:
The `MSSQLQueryBuilder` only needs to the default schema and not the entire MSSQLConfiguration object thats we replaced it with the default schema

gosusnp reviewed (commented):
I focused mostly on the SQL integration piece as I have little context on the blob store part.
The structure feels fine but I do have a couple of concern
* looks like we added yet another type structure for CSV conversion
* SQL for the bulk feels duplicated from the initial part
* Escaping is different and relies on split of built string

I don't have a strong opinion against the fork, but I think it makes maintenance harder because of the code duplication. I highly suspect it would be very easy for us to fix a typing issue in one and overlook the same issue in the other.
@subodh1810, WDYT?

subodh1810 commented on airbyte-integrations/connectors/destination-mssql-v2/src/main/kotlin/io/airbyte/integrations/destination/mssql/v2/MSSQLChecker.kt:
```diff
@@ -27,19 +43,122 @@ class MSSQLChecker(private val dataSourceFactory: MSSQLDataSourceFactory) :
+                if (
+                    config.mssqlLoadTypeConfiguration.loadTypeConfiguration is BulkLoadConfiguration
+                ) {
+                    val bulkLoadConfig = config.mssqlLoadTypeConfiguration.loadTypeConfiguration
```
fixed

subodh1810 commented:
> I focused mostly on the SQL integration piece as I have little context on the blob store part. The structure feels fine but I do have a couple of concern
> 
> * looks like we added yet another type structure for CSV conversion
> * SQL for the bulk feels duplicated from the initial part
> * Escaping is different and relies on split of built string
> 
> I don't have a strong opinion against the fork, but I think it makes maintenance harder because of the code duplication. I highly suspect it would be very easy for us to fix a typing issue in one and overlook the same issue in the other. @subodh1810, WDYT?

@gosusnp I addressed all your comments except for reusing the SQL query! I think we could re-use the 2 queries (MERGE query and Schema fetch) query. I will create a ticket to refactor it and do it as part of a separate PR. The schema fetch query would require some refactoring in the schema update logic as well and I want to avoid doing that as part of this PR
